### ISSUE TITLE
Add support for Input Methods

### ISSUE THREAD

**Issue by JordanMartinez** [FXMisc/RichTextFX#146]
Hello Tomas,
What would be the best way to incorporate multiple language inputs? What would be the best way to handle this?

I think languages like Spanish or Greek wouldn't be too hard to do. One could remap the KeyTyped events to insert the correct letter in their respective alphabets.

For other languages (like Chinese using PinYin or some other input method), this wouldn't work. For my purposes, I don't need support for a more complex language, just the simpler ones. However, finding a way to incorporate the system's input method would prevent each developer from having to create 1) a separate KeyTyped events remapping for each language used and 2) a method to switch from one keyboard layout to another, depending on what the user needs.

**Comment by TomasMikula** [comment]
Hi Jordan,
different keyboard layouts, like Spanish or Greek, should just work. Is that not your experience? JavaFX's KeyTyped event already contains a keyboard layout-dependent character. Input methods and right-to-left (or bidirectional) text is not supported.

**Comment by JordanMartinez** [comment]
Sorry, I only meant input methods. I'm using a US keyboard layout, but if I wanted to quickly hop over to a Spanish/Greek/Chinese input method, it won't register.

**Comment by TomasMikula** [comment]
I don't have any experience using input methods. A good place to start would be to look at what JavaFX's TextArea does.

**Comment by JordanMartinez** [comment]
I looked into TextArea and input methods in general. Here's what I found.

First, there's a specific Event for this: [InputMethodEvent](https://docs.oracle.com/javase/8/javafx/api/javafx/scene/input/InputMethodEvent.html). 
Second, the API says that the event is delivered to the Node that extends TextInputControl, which TextArea extends and which RichTextFX does not extend (it extends Control). I'm not sure what the underlying code is in TextInputControl.

**Comment by JordanMartinez** [comment]
The problem is that an area's InputMethodRequests value is null. Any InputMethodEvent won't be registered until this gets taken care of. My workaround is to "steal" a TextArea's InputMethodRequests (an instance of TextInputControlSkin). However, I don't fully understand the implementation of this (as you can see below). So, I hope you have more insight as to how to do this better. :-) 

(Also, I know this isn't Java, but you can still see what I'm doing. I apologize in advance!)

````groovy
import static groovyx.javafx.GroovyFX.start
start {
    stage(id: "stage", title: "Hello", primary: true, visible: true) {
        scene(width: 500, height: 500) {
            vbox(id: "vbox") {
                textArea(id: "textArea")
            }
        }
    }
    def richArea = new InlineCssTextArea()

    // richArea's InputMethodRequests is null.
    assert richArea.getInputMethodRequests() == null

    // "Stealing" the InputMethodRequests from a TextArea will work. However,
    //  it must already be initialized:

    // Won't work, and I'm not sure why. It's probably an issue with using GroovyFX...
    def badArea = new TextArea()
    vbox.getChildren().add(badArea)
    assert badArea.getInputMethodRequests() == null
    vbox.getChildren().remove(badArea)

    // Will work, but I'm not sure why it does rather than 'badArea'
    // textArea's InputMethodRequests is an instance of TextInputControlSkin
    assert textArea.getInputMethodRequests() != null
    richArea.setInputMethodRequests(textArea.getInputMethodRequests())

    // remove unneeded textArea and add richArea
    vbox.getChildren().remove(textArea)
    vbox.getChildren().add(richArea)

    // Now that richArea has an InputMethodRequests value, InputMethodEvents
    //  will register if a handler/filter is used. It won't without it.

    // NOTE: This handler only inserts the finished text. It doesn't insert the
    //  text that a user is still 'composing,' and replace this 'composed' text with the finished text.
    richArea.setOnInputMethodTextChanged({ e ->
        if (e.committed != "") {
            richArea.insertText(richArea.caretPosition, e.committed)
        }
    })
}
````

**Comment by TomasMikula** [comment]
Thanks for posting back your workaround!

**Comment by JordanMartinez** [comment]
No problem!

**Comment by JordanMartinez** [comment]
I'm reopening this to add official support for this out-of-box.

**Comment by oldsin** [comment]
As a Chinese user, I guess people who use the most commonly used input method of China, Pinyin, would like to see the candidate window (for picking possible words) just below the caret.

Below is a sample input of Chinese characters “你好” (Greetings) using Pinyin in macOS Pages, the candidate window appears just below the caret and before the first input character (which is also the initial caret position).

![chinese_input_using_pages](https://user-images.githubusercontent.com/16722967/61593587-0da8cb00-ac14-11e9-8154-6f51db57665f.png)

The workaround above works pretty well! But it doesn't show the candidate window in the place where the candidate window should be. To fix this, @JordanMartinez points out that `Caret#getCaretBounds()` can be used to get the caret's position on screen in #753. By passing the  rich text area to the `InputMethodRequestsObject`, the `InputMethodRequestsObject#getTextLocation(int)` can be modified to show the effect presented above:
```java
    @Override
    public Point2D getTextLocation(int offset) {
        // a very rough example, only tested under macOS
        Optional<Bounds> caretPositionBounds = area.getCaretBounds();
        if (caretPositionBounds.isPresent()) {
            Bounds bounds = caretPositionBounds.get();
            return new Point2D(bounds.getMaxX() - 5, bounds.getMaxY());
        }

        throw new NullPointerException();
    }
```
![chinse_input_using_generic_styled_area](https://user-images.githubusercontent.com/16722967/61593834-48f8c900-ac17-11e9-956e-23a92b308cb2.png)

**Comment by xulihang** [comment]
Hi there,

I tested the workaround and it works. Great thanks.

I modified the code a bit to replace selected text with text commited
```
richArea.setOnInputMethodTextChanged({ e ->
        if (e.committed != "") {
            richArea.replaceSelection("")
            richArea.insertText(richArea.caretPosition, e.committed)
        }
    })
```

As for the candidate window, I found that only macOS can display it correctly. Native TextArea has this problem as well. The candidate window is at the lower-right under Windows 10 and is at the lower-left area of the application under Linux. I tested on JRE 1.8.
